### PR TITLE
Added Morvan's ddpg.py algorithm

### PR DESCRIPTION
**Pull request by wildomonges**

### CODE REVIEW

File: src/motion/nodes/DDPG.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+DDPG is Actor Critic based algorithm.
+Pendulum example.
+View more on my tutorial page: https://morvanzhou.github.io/tutorials/
+Using:
+tensorflow 1.0
+gym 0.8.0
+"""
+
+import tensorflow as tf
+import numpy as np
```
Comment (void-robotics) [inline src/motion/nodes/DDPG.py]: although np is standard for numpy, I don't think it's friendly to new people. I think we should have it be `import numpy as math` or something

Comment (void-robotics) [reply src/motion/nodes/DDPG.py]: same for tf

File: src/motion/nodes/DDPG.py [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+
+        with tf.control_dependencies(target_update):
+            # soft replacement happened at here
+            q_target = self.reward + self.gamma * q_network_next
+            td_error = tf.losses.mean_squared_error(labels=q_target, predictions=q_network)
+            self.critic_train = tf.train.AdamOptimizer(self.lr_c).minimize(td_error, var_list=critic_params)
+
+        self.sess.run(tf.global_variables_initializer())
+
+    def choose_action(self, state):
```
Comment (void-robotics) [inline src/motion/nodes/DDPG.py]: every "def" function should have comments like C++ (see CO 3); see Coding Conventions. Create a conventions like C++ but with the Python syntax and update the Coding Conventions

File: src/motion/nodes/DDPG.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+        :return: tensorflow session action
+        """
+        return self.sess.run(self.action, {self.state: state[np.newaxis, :]})[0]
+
+    def learn(self):
+        """
+        Lean the model
+        """
+        indices = np.random.choice(self.memory_capacity, size=self.batch_size)
+        bt = self.memory[indices, :]
```
Comment (void-robotics) [inline src/motion/nodes/DDPG.py]: see VA 4

File: src/motion/nodes/DDPG.py [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+        # soft update operation
+        target_update = [ema.apply(action_params), ema.apply(critic_params)]
+        # replaced target parameters
+        next_action = self._build_a(self.next_state, reuse=True, custom_getter=ema_getter)
+        q_network_next = self._build_c(self.next_state, next_action, reuse=True, custom_getter=ema_getter)
+        # maximize the q
+        action_loss = - tf.reduce_mean(q_network)
+        self.action_train = tf.train.AdamOptimizer(self.lr_a).minimize(action_loss, var_list=action_params)
+
+        with tf.control_dependencies(target_update):
```
Comment (void-robotics) [inline src/motion/nodes/DDPG.py]: see CO 2

File: src/motion/nodes/DDPG.py [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+            return tf.layers.dense(net, 1, trainable=trainable)  # Q(s,a)
+
+
+""" training """
+
+env = gym.make(ENV_NAME)
+env = env.unwrapped
+env.seed(1)
+
+state_dim = env.observation_space.shape[0]
```
Comment (void-robotics) [inline src/motion/nodes/DDPG.py]: see CO 2

File: src/motion/nodes/DDPG.py [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+action_dim = env.action_space.shape[0]
+action_bound = env.action_space.high
+
+ddpg = DDPG(action_dim, state_dim, action_bound)
```
Comment (void-robotics) [inline src/motion/nodes/DDPG.py]: see CO 2

File: src/motion/nodes/DDPG.py [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+        if RENDER:
+            env.render()
+
+        # Add exploration noise
+        action = ddpg.choose_action(state)
+        # add randomness to action selection for exploration
+        action = np.clip(np.random.normal(action, control_exploration), -2, 2)
+        next_state, reward, done, info = env.step(action)
+
+        ddpg.store_transition(state, action, reward / 10, next_state)
```
Comment (void-robotics) [inline src/motion/nodes/DDPG.py]: see CO 2

File: src/motion/nodes/DDPG.py [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+
+        if ddpg.pointer > MEMORY_CAPACITY:
```
Comment (void-robotics) [inline src/motion/nodes/DDPG.py]: see CO 2

File: src/motion/nodes/DDPG.py [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+            # decay the action randomness
+            control_exploration *= .9995
+            ddpg.learn()
+
+        state = next_state
```
Comment (void-robotics) [inline src/motion/nodes/DDPG.py]: see CO 2

**Review by ghost — CHANGES_REQUESTED** [review]

**Review by ghost — APPROVED** [review]

**Pull request merged**